### PR TITLE
use more braces to avoid `where T = ...` syntax

### PR DESCRIPTION
**Pull request by JeffBezanson**

### CODE REVIEW

**Review by tkelman — APPROVED** [review]
thank you

File: base/boot.jl [hunk 1]
```diff
@@ -269,9 +269,9 @@ Void() = nothing
 
 immutable VecElement{T}
     value::T
-    VecElement{T}(value::T) where T = new(value) # disable converting constructor in Core
+    VecElement{T}(value::T) where {T} = new(value) # disable converting constructor in Core
 end
-VecElement(arg::T) where T = VecElement{T}(arg)
+VecElement(arg::T) where {T} = VecElement{T}(arg)
```
Comment (stevengj) [inline base/boot.jl]: For simple cases like this, I find the older syntax `VecElement{T}(arg::T) = VecElement{T}(arg)` to be much clearer than `where {T}`.   Why do we need to use `where` here?

Comment (JeffBezanson) [inline base/boot.jl]: Best demonstrated by example. What would

```
Complex{Float64}(x::Float64) = ...
```

mean?

Also note the old syntax is very close to

```
Vec{T}(arg) = Vec{T}(arg)
```

which looks circular. The `{T}` on the left means something very different than the `{T}` on the right.

Comment (stevengj) [inline base/boot.jl]: Presumably `Complex{Float64}(x::Float64) = ...` means the same as `Complex{T}(x::T) = ...` the `T` (whatever it is called) is a local variable that overrides any definition of `T` in the enclosing scope.

I don't recall people complaining about either of these things before.   `where` is great when you need triangular dispatch etc, but the old syntax still seems nice to me as an abbreviation.

Comment (JeffBezanson) [inline base/boot.jl]: It's a pretty common complaint, e.g. looking at the comments in #11310.

Currently the syntax

```
function Complex{T}(x::T) where T
```

unambiguously defines a constructor for `Complex{T}`. But if you remove the `where T`, the `{T}` is interpreted as introducing static parameters, so instead it defines a constructor for `Complex`. This is confusing. Alternatives like `Complex{T}{T}(x)` or `Complex{T}(x){T}` seem more obscure to me.

Comment (Sacha0) [inline base/boot.jl]: > Presumably Complex{Float64}(x::Float64) = ... means the same as Complex{T}(x::T) = ... the T (whatever it is called) is a local variable that overrides any definition of T in the enclosing scope.

`Complex{Float64}(x::Float64) = ...` defines a constructor for `Complex{Float64}`, no? Whereas `Complex{T}(x::T) = ...` defines a parameterized method family for `Complex`?

Also ref. https://github.com/JuliaLang/julia/pull/20444. Best!

**Comment by kmsquire** [comment]
> I don't recall people complaining about either of these things before. 

I don't remember where, but FWIW I know that I commented a while ago about the confusion between parametric functions and parametric type constructors (especially external "constructors", as in Jeffs example above).

I think that was one of the more confusing syntax choices for beginners, and even though I wasn't in love with "where", I'm very glad this bit of syntax overlap is going away.

**Pull request merged**